client-core/src/wallet/syncer.rs

```diff
@@ -106,6 +106,8 @@ pub struct SyncerConfig<S: SecureStorage, C: Client> {
     block_height_ensure: u64,
 }
 
+pub type SyncCallback = extern "C" fn(f32) -> i32;
```

Comment:
The callback probably should be a closure to be useful, which can encapsulate the context for the c function.
Also if we choose callback to report the progress, ideally, should deprecate the `progress_reporter`, which can be implemented based on the callback feature.

Reply:
yes, but this callback is "C" function

Reply:
I think a c function together with a context pointer can be wrapped in a rust closure?

Reply:
that's possible

Reply:
> that's possible
> but i don't need additonal thread for this.
> what do you think?
> 
> if sync is persistent, current progress updater is fine.
> but for c-calling, it's just one thread for both of producer, consumer
> 
> thread context is very important in c, because other c-source code will be thread-unsafe.

No thread, just callbacks. 
By context, I mean the c callback function needs something to access the state, otherwise, all it have is an integer argument, what can it do other than printing it out. It needs something to connect back to the system.

Reply:
got it,
callback is pure event info, just plain c primitive type would be fine.
don't need any opaque pointer

because it is just giving information from core to sdk-user.

Reply:
Use closure in logic code:
```
callback: F,
where F: Fn(int, int) -> bool
```
The ffi detail should be wrapped in the api library.

Ideally, we should implement progress_report with callback, but that can be done in another PR I think.